dask/dataframe/_hyperloglog.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*
+u"""Implementation of HyperLogLog
+
+This implements the HyperLogLog algorithm for cardinality estimation, found
+in
+
+    Philippe Flajolet, Éric Fusy, Olivier Gandouet and Frédéric Meunier.
+        "HyperLogLog: the analysis of a near-optimal cardinality estimation
+        algorithm". 2007 Conference on Analysis of Algorithms. Nice, France
+        (2007)
+
+"""
+from __future__ import division
+
+import numpy as np
+import pandas as pd
+
+def compute_first_bit(a):
+    # Consider making this less memory-hungry
+    #bits = np.bitwise_and.outer(a, 1<<np.arange(31, -1, -1))
+    bits = np.bitwise_and.outer(a, 1<<np.arange(32))
+    bits = bits.cumsum(axis=1).astype(np.bool)
+    return 33 - bits.sum(axis=1)
+
+def terrible_dataframe_hash(df):
```

Comment:
That's a fair point, this hasn't come up before. `hashlib` could be used to do consistent hashing across sessions, but wouldn't work for any object (since object arrays aren't necessarily strings). `hash` is also much faster than the methods found in `hashlib`. Consistent hashseeds are used between processes in `multiprocessing.Pool`, so if you're not persisting hashes between sessions then this shouldn't be a problem:

```python
# python 3.5, no PYTHONHASHSEED set
In [1]: import multiprocessing as mp

In [2]: p = mp.Pool()

In [3]: data = ['abcde' for i in range(1000)]

In [4]: h = p.map(hash, data)

In [5]: set(h)
Out[5]: {4002444540688553378}
```

A consistent hashseed between `dask-worker`s would probably be a good idea though.